packages/cli/src/commands/hydrogen/link.ts

```diff
@@ -100,18 +119,94 @@ export async function linkStorefront({
       label: `${title} (${productionUrl})`,
     }));
 
+    choices.unshift({
+      value: CREATE_NEW_STOREFRONT_ID,
+      label: 'Create a new storefront',
+    });
+
     const storefrontId = await renderSelectPrompt({
       message: 'Choose a Hydrogen storefront to link',
       choices,
     });
 
-    selectedStorefront = storefronts.find(({id}) => id === storefrontId);
+    if (storefrontId === CREATE_NEW_STOREFRONT_ID) {
+      selectCreateNewStorefront = true;
+    } else {
+      selectedStorefront = storefronts.find(({id}) => id === storefrontId);
+    }
   }
 
-  if (!selectedStorefront) {
-    return;
+  if (selectCreateNewStorefront) {
+    const storefront = await createNewStorefront(path, shop);
+
+    if (!storefront) {
+      return;
+    }
+
+    selectedStorefront = storefront;
   }
 
+  if (selectedStorefront) {
+    await linkExistingStorefront(path, selectedStorefront, silent, cliCommand);
+  }
+}
+
+async function createNewStorefront(
+  path: string | undefined,
+  shop: string,
+): Promise<HydrogenStorefront | undefined> {
+  const projectDirectory = path && basename(path);
+
+  const projectName = await renderTextPrompt({
+    message: 'What do you want to name the Hydrogen storefront on Shopify?',
+    defaultValue: projectDirectory || 'Hydrogen Storefront',
+  });
+
+  let storefront: HydrogenStorefront | undefined;
+  let jobId: string | undefined;
+
+  await renderTasks([
+    {
+      title: 'Creating storefront',
+      task: async () => {
+        const result = await createStorefront(shop, projectName);
+
+        storefront = result.storefront;
+        jobId = result.jobId;
+
+        if (result.userErrors.length > 0) {
+          renderUserErrors(result.userErrors);
+        }
+      },
+    },
+    {
+      title: 'Creating API tokens',
+      task: async () => {
+        try {
+          await waitForJob(shop, jobId!);
+        } catch (_error) {
+          renderWarning({
+            headline: UNKNOWN_ERROR_HEADLINE,
+            body: [
+              'Please try again or contact support if the error persists.',
+            ],
+          });
+          storefront = undefined;
+        }
```

Comment:
The state on Admin:
- When the job fails on Admin, it will perform cleanup on the Storefront

The state in your local environment:
- No storefront is returned, so you aren't linked with anything

> Is there anything the user need to do in the Admin panel to fix it?
I don't think so. When the job fails, it's something on our end, that's why i give a generic error message in the body. The previous step in this `renderTasks` relayed the userError to the user because the user _can_ fix those issues (e.g. title too long, title too short, limit on storefronts, etc.)